test/jdk/javax/swing/JFileChooser/FileSystemView/SystemIconTest.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8182043
+ * @summary Access to Windows Large Icons
+ * sun.awt.shell.ShellFolder
+ * @run main SystemIconTest
+ */
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.filechooser.FileSystemView;
+import java.awt.image.MultiResolutionImage;
+import java.io.File;
+
+public class SystemIconTest {
+    static final FileSystemView fsv = FileSystemView.getFileSystemView();
+
+    public static void main(String[] args) {
+        testSystemIcon();
+        negativeTests();
+    }
+
+    static void testSystemIcon() {
+        String os = System.getProperty("os.name");
+        if (os.startsWith("Windows")) {
+            String windir = System.getenv("windir");
+            testSystemIcon(new File(windir), true);
+            testSystemIcon(new File(windir + "/explorer.exe"),
+                    true);
+        } else {
+            String homedir = System.getProperty("user.home");
+            testSystemIcon(new File(homedir), false);
+        }
+    }
+
+    static void negativeTests() {
+        Icon icon;
+        try {
+            icon = fsv.getSystemIcon(new File("."), -1, 16);
+            throw new RuntimeException("Negative size icon should throw invalid argument exception");
+        } catch (IllegalArgumentException iae) {
+            // Expected
+        }
+
+        icon = fsv.getSystemIcon(new File("thereisdefinitelynosuchfile.why"),
+                16, 16);
+        if (icon != null) {
+            throw new RuntimeException("Icons for files with invalid names should be null");
+        }
+    }
+
+    static void testSystemIcon(File file, boolean implComplete) {
+        int[] sizes = new int[] {16, 32, 48, 64, 128};
+        for (int size : sizes) {
+            ImageIcon icon = (ImageIcon) fsv.getSystemIcon(file, size, size);
```

Comment:
Is this cast allowed without instanceof check?

Reply:
On Windows - yes. When implementation is complete on other platforms test will have to be updated depending on the way it will be implemented there.

Reply:
Hmm .. I focused too much on the spec it seems. The API returns Icon and the test should not
"know" or depend on it being anything more specific.
Why does it need to do this ?